### DISCUSSION TITLE
Solid and Mastodon

### DISCUSSION THREAD

**Discussion by dj-sf** [mastodon/mastodon#26403] [General]
Solid is an initiative that aims to decouple user data from applications. It is still quite new, but it has some high profile supporters (including Tim Berners-Lee, who created it).

Although it's distinct from the Fediverse (in that it has no direct relationship with ActivityPub), it supports the goal of internet decentralization and has features that could have a lot to offer Fediverse projects.  For example, it could elegantly address some [long-standing issues around instance migration](https://github.com/mastodon/mastodon/issues/12423) by decoupling user data from the instances themselves.

[Here is a link to the Solid project website](https://solidproject.org/). Do you think Mastodon (or the Fediverse in general) could benefit from Solid?  I'd love to hear your thoughts!

**Comment by dj-sf** [comment]
Below is a rough diagram of what I am proposing i.e. how I think Solid could be used to improve Mastodon (and other Fediverse projects).  Would love to hear what ppl think:

![Mastodon-with-Solid-Rough Draft Diagrams](https://github.com/mastodon/mastodon/assets/30304653/3dc381f7-c409-48f6-bd9c-33db98d3b709)

**Reply by dj-sf** [reply]
I have created a fork named [mastodon-solid-poc](https://github.com/dj-sf/mastodon-solid-poc) to begin working on a proof-of-concept.  Please feel free to reach out if you want to discuss or contribute!

**Comment by dj-sf** [comment]
Just FYI I have also started [a discussion in the Solid forum](https://forum.solidproject.org/t/solid-and-mastodon/6837) about this topic.

**Reply by hzbarcea** [reply]
Hi @dj-sf , interesting project. I am curious about your project and briefly looked at the code.

Btw, it looks like a perfect fit for the [Solid Hackathon](https://www.inrupt.com/event/solid-hackathon/home) that will start on 09/11. Since you mentioned TimBL he would like to see such [integrations](https://matrix.to/#/!BBIXfVcLmmoJwIljrW:gitter.im/$1Uv7-Lv9mCbmYI0wK9eiS1TWIdkAeLrJMYQ-TZZCJAY) (with Mastodon, Slack and others). I would strongly encourage you to participate to the hackathon with this project. 

I'd like to discuss more, if you'd be interested.

**Reply by dj-sf** [reply]
This is a great idea! wonderful coincidence that this hackathon is happening now of all times.  Would love to sync up and chat more.  Looks like signup is open for the hackathon until September 11, so I'm gonna try to get a team together by then!

The repo i linked is still pretty boring because i haven't really started yet 😂 but this hackathon is gonna change that

**Reply by dj-sf** [reply]
@hzbarcea quick update -- I have submitted the project proposal for the Solid Hackathon via the link you provided.  I do have a few questions though so if you are still open to discussing further please shoot me a DM on my Mastodon account [@djsf@fosstodon.org](https://fosstodon.org/@djsf) and we can go from there.  Or if you would prefer a different method of getting in touch please let me know.  Thanks! :)

**Reply by hzbarcea** [reply]
Will do @dj-sf

**Comment by csarven** [comment]
The fundamental alignment is about [ActivityPub](https://www.w3.org/TR/activitypub/) and [Solid Protocol](https://solidproject.org/TR/protocol) (and possibly other specifications under https://solidproject.org/TR/ ) , and the degree in which Mastodon can conform to those specifications, in particular the applicable classes of products. Mastodon already does a great job with AP - we know this from when Mastodon's implementation of AP back when we worked on the Social Web WG.  [Linked Data Notifications](https://www.w3.org/TR/ldn/) is one particular specification that's used by both ActivityPub and Solid Protocol. See also https://github.com/solid-contrib/external-interop-panel/issues/4#issuecomment-640227867 for a lot of details and links on past work.

The good news is that some folks in both [W3C Social Web Community Group](https://www.w3.org/groups/cg/swicg) and the [W3C Solid Community Group](https://www.w3.org/groups/cg/solid/) are interested in better identifying and bridging the gaps.

Folks are welcome to join either or both Community Groups to move things forward, and there are already implementations that's conforming to some of the specifications mentioned above (e.g., https://github.com/assemblee-virtuelle/activitypods , https://github.com/linkeddata/dokieli ). I'd caution against building anything towards proprietary solutions by private dev gatherings, and stick to open standards development communities.

On 2023-09-11, the Solid CG has an event at W3C TPAC: https://www.w3.org/events/meetings/3d93d256-f017-48c6-a509-9bd089a714e3/ (see link for call details and agenda WIP) where anyone is welcome to join.

Yet another event is at SWIB23, https://swib.org/swib23/programme.html#keynote-inherently-social-decentralized-and-for-everyone , where I will give a keynote quite literally on this topic :) I think a recording will be available later.

>decipher the underlying philosophies and technical architectures behind initiatives like the Solid project and the Fediverse, [..] analyze the possibilities for interoperability among different classes of products

**Reply by dj-sf** [reply]
Thanks for sharing all these! Been a busy week but I'll check these out this weekend and get back to you 🤙

**Comment by outlaw-dame** [comment]
This is very exciting to see there’s significant interest in brining the Fediverse and Solid together. I know some devs that have experimented with Mastodon and Solid.
We had thought about the same but decided to make another project altogether and we came across ActivityPods.
We are in the process of building a Fediverse platform using ActivityPods.

**Comment by dj-sf** [comment]
Hi everyone, wanted to provide a brief update on this.

I did a lot of work over the last month on figuring out how to make this work, and I still believe it could work well.  I tried implementing it as part of a hackathon and got some clarity about how exactly it should be implemented, but was not able to get it fully implemented by the deadline due to my poor understanding OAuth 2.0 and OIDC. 

Took a little breather over the last couple of weeks for personal reasons but I'm getting back into this.  Currently studying OAuth 2.0 and OIDC to assist with solving this problem.  

Stay posted for more updates over the next few weeks!

**Reply by erlend-sh** [reply]
I've made a refreshed [case for Solid + ActivityPub interoperability here](https://socialhub.activitypub.rocks/t/how-solid-and-activitypub-complement-each-other-best/727/14?u=erlend_sh).

**Comment by roke-julian-lockhart** [comment]
https://ruben.verborgh.org/blog/2022/12/30/lets-talk-about-pods/#document-centric
